### PR TITLE
CD-18: iPerf Speed test results

### PR DESCRIPTION
**Pull request by mariemm97**

### CODE REVIEW

File: examples/ssh/iperf.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/**
+ * this driver is designed to retrieve network speed data using the iperf3.
+ * Communication protocol is SSH.
+ * This driver create a dynamic monitoring variables 
+ * down: test for download speed.
+ * up: test for upload speed.
+ * UDPdown: test for UDP download speed.
+ * UDPup: test for UDP upload speed.
```
Comment (mohamedkallel) [inline examples/ssh/iperf.js:8]: add the iperf version tested with and the operating system used

**Review by mohamedkallel — CHANGES_REQUESTED** [review]

**Review by mohamedkallel — APPROVED** [review]

File: examples/ssh/iperf.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ */
+
+// Define SSH configuration
+var sshConfig = {
+    port: 22,
+    timeout: 20000
+};
+
+var _var = D.device.createVariable;
+var dload, uload, udpdload, udpuload;  // Define variables for storing download/upload speeds
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:20]: use more descriptive names
var downloadSpeed, uploadSpeed, uploadSpeedUDP, downloadSpeedUDP

File: examples/ssh/iperf.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ * Tested with iperf3 version: v 3.7 under Ubuntu 22.04.1 LTS
+ */
+
+// Define SSH configuration
+var sshConfig = {
+    port: 22,
+    timeout: 20000
+};
+
+var _var = D.device.createVariable;
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:19]: use D.createVariable directly and avoid defining this

File: examples/ssh/iperf.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+var dload, uload, udpdload, udpuload;  // Define variables for storing download/upload speeds
+
+var UDP = "Y"; // Define whether UDP is enabled on the device (set to "Y" by default)
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:22]: What is this variable for?  And if its conditional why not set it a boolean? 
If its a flag to test UDP or not name it properly

eg 'shouldTestUDPSpeed'

File: examples/ssh/iperf.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+function checkSshError(err) {
+    if (err.message) console.error(err.message);
+    if (err.code == 5) D.failure(D.errorType.AUTHENTICATION_ERROR);
+    if (err.code == 255) D.failure(D.errorType.RESOURCE_UNAVAILABLE);
+    D.failure(D.errorType.GENERIC_ERROR);
+}
+
+// Function for executing SSH command 
+function exec_command(command, callback) {
+    var config = JSON.parse(JSON.stringify(sshConfig));
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:42]: Why do you stringify and then parse?

File: examples/ssh/iperf.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+//Checking SSH errors and handling them
+function checkSshError(err) {
+    if (err.message) console.error(err.message);
+    if (err.code == 5) D.failure(D.errorType.AUTHENTICATION_ERROR);
+    if (err.code == 255) D.failure(D.errorType.RESOURCE_UNAVAILABLE);
+    D.failure(D.errorType.GENERIC_ERROR);
+}
+
+// Function for executing SSH command 
+function exec_command(command, callback) {
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:41]: use camelCase

File: examples/ssh/iperf.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    config.command = command;
+    D.device.sendSSHCommand(config, function (out, err) {
+        if (err) checkSshError(err);
+        callback(out.split("\n"));
+    });
+}
+
+//This function execute the SSH commands to retrieve network speed data using the iperf3 tool, and then creating dynamic monitoring variables based on the obtained data.
+function execute() {
+    exec_command(commands[0], function (cmd) {
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:52]: Just use promises to execute all of the commands 

the callback structure is a bit strange for cascading 4 calls

File: examples/ssh/iperf.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        dload = cmd || 0;
+        exec_command(commands[1], function (cmd) {
+            uload = cmd || 0;
+            if (UDP == "Y") {
+                exec_command(commands[2], function (cmd) {
+                    udpdload = cmd || 0;
+                    exec_command(commands[3], function (cmd) {
+                        udpuload = cmd || 0;
+                        D.success([
+                            _var("down", "DLOAD", dload, "Mb/s"),
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:62]: use better naming and uids

download-speed
upload-speed
download-speed-udp
download-speed-UDP

File: examples/ssh/iperf.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                            _var("up", "ULOAD", uload, "Mb/s"),
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:63]: The Labels should be clear as well

"Upload Speed" etc

File: examples/ssh/iperf.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+* @remote_procedure
+* @label Validate Association
+* @documentation This procedure is used to validate if iperf3 commands are executed successfully
+*/
+function validate() {
+    execute(function () {
+        D.success()
+            .catch(failure);
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:93]: the catch must be for the execution function not the D.success

File: examples/ssh/iperf.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+* @remote_procedure
+* @label Get Device Variables
+* @documentation This procedure is used for retrieving iperf3 results
+*/
+function get_status() {
+    execute(function () {
+        D.success()
+            .catch(failure);
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:105]: the catch must be for the execute function

File: examples/ssh/iperf.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    console.log(err);
+    D.failure(D.errorType.GENERIC_ERROR);
+}
+
+/**
+* @remote_procedure
+* @label Validate Association
+* @documentation This procedure is used to validate if iperf3 commands are executed successfully
+*/
+function validate() {
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:90]: This validate function does not capture erroneous cases

Eg validate on any device with a working SSH will result in an OK and it MUST NOT

Please add a proper validation if the device has iperf3 installed and can run the respective commands

File: examples/ssh/iperf.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+                            _var("UDPup", "UDPULOAD", udpuload, "Mb/s")
+                        ]);
+                    });
+                });
+            } else {
+                D.success([
+                    _var("down", "DLOAD", dload, "Mb/s"),
+                    _var("up", "ULOAD", uload, "Mb/s")
+                ]);
+            }
```
Comment (GeorgiTsatsev) [inline examples/ssh/iperf.js:74]: What about a failure in parsing (eg no output)???

**Review by mohamedkallel — APPROVED** [review]

**Review by GeorgiTsatsev — APPROVED** [review]

**Pull request merged**